### PR TITLE
Wrap legacy APIs

### LINKED ISSUE #464: Transition away from positional APIs
We never wanted APIs that can be used with more that ~2 positional parameters. We should go to keyword-only-parameters.

This can be done via [legacy-api-wrap](https://github.com/flying-sheep/legacy-api-wrap)

### PR DESCRIPTION
**Pull request by flying-sheep**
Part of [scanpy 2.0](https://github.com/orgs/scverse/projects/18/views/1)

- [ ] Closes #
- [x] Tests included or not required because:

- [ ] Release notes not necessary because:

I implemented a Ruff check (PLR0917) for this, but setting `max-positional-args = 3` would make this massive PR even larger, so I opted to do it in a separate one.

## Reviewers

Your main job is to check if the position of the `*` makes sense for each exported function (i.e. the ones with the `@legacy_api` decorator). I tried my best to base it on internal usage of each API, but one placement or the other might be to early.

The only real logic changes are in `scanpy/tests/test_package_structure.py`. This PR:

- makes public APIs with more than 5 parameters keyword-only without breaking backwards compatibility (for now)
- makes private APIs with more than 5 parameters keyword-only
- checks that we don’t internally use the old positional APIs using a warning filter with `action='error'`
- checks that APIs use our convention for the `copy` parameter (`adata` as first param, returns `adata` type or None`)
- manually checked that APIs use our convention `filename: Path | str`/`path: Path | str`

## Follow-up changes

- [ ] Set `max-positional-args = 3`
- [ ] make sure that no new public API gets introduced without being included in the `api_module_names` list
- [ ] enforce convention `show`, `return_fig`, `ax`
- [ ] enforce convention for `random_seed: AnyRandom`
- [x] https://github.com/scverse/scanpy/issues/2331

### CODE REVIEW

File: scanpy/external/pp/_bbknn.py [hunk 1]
```diff
@@ -70,18 +73,17 @@ def bbknn(
...
-        'rogerstanimoto', 'russellrao', 'sokalsneath', 'sokalmichener', 'yule'])
+        'linfty', 'linf', 'minkowski', 'seuclidean', 'standardised_euclidean', 'wminkowski', ...])
 
         KDTree supports members of :class:`sklearn.neighbors.KDTree`’s ``valid_metrics`` list, or parameterised
         :class:`~sklearn.metrics.DistanceMetric` objects:
 
+        >>> import sklearn.neighbors
         >>> sklearn.neighbors.KDTree.valid_metrics
-        ['p', 'chebyshev', 'cityblock', 'minkowski', 'infinity', 'l2', 'euclidean', 'manhattan', 'l1']
+        ['euclidean', 'l2', 'minkowski', 'p', 'manhattan', 'cityblock', 'l1', 'chebyshev', 'infinity']
```
Comment (ilan-gold) [inline scanpy/external/pp/_bbknn.py]: Why did this change?

File: scanpy/external/pp/_hashsolo.py [hunk 2]
```diff
@@ -261,15 +264,17 @@ def _calculate_bayes_rule(data, priors, number_of_noise_barcodes):
...
 def hashsolo(
-    adata: anndata.AnnData,
-    cell_hashing_columns: list,
-    priors: list = [0.01, 0.8, 0.19],
-    pre_existing_clusters: str = None,
-    number_of_noise_barcodes: int = None,
+    adata: AnnData,
+    cell_hashing_columns: Sequence[str],
+    *,
+    priors: tuple[float, float, float] = (0.01, 0.8, 0.19),
```
Comment (ilan-gold) [inline scanpy/external/pp/_hashsolo.py:275]: Was the type annotation here wrong?

Comment (ilan-gold) [reply scanpy/external/pp/_hashsolo.py:275]: Or is this just because lists and tuples are somewhat polymorphic?

File: scanpy/external/tl/_phenograph.py [hunk 3]
```diff
@@ -18,10 +19,29 @@
...
+    "resolution_parameter",
+    "n_iterations",
+    "use_weights",
+    "seed",
+    "copy",
+)
 @doctest_needs("phenograph")
 def phenograph(
-    adata: AnnData | np.ndarray | spmatrix,
+    data: AnnData | np.ndarray | spmatrix,
```
Comment (ilan-gold) [inline scanpy/external/tl/_phenograph.py:44]: Isn't this a breaking a change?

File: scanpy/external/tl/_phenograph.py [hunk 4]
```diff
@@ -246,9 +265,10 @@ def phenograph(
...
     if copy:
         return communities, graph, Q
-    else:
-        adata.obsp[ig_key] = graph.tocsr()
-        if comm_key:
-            adata.obs[comm_key] = pd.Categorical(communities)
-        if Q:
-            adata.uns[q_key] = Q
+
+    assert isinstance(data, AnnData)
```
Comment (ilan-gold) [inline scanpy/external/tl/_phenograph.py]: I get the impulse to test this here, but seeing this in this PR makes me think something might be broken that is not stated here and is somehow being fixed by this.

File: scanpy/neighbors/__init__.py [hunk 5]
```diff
@@ -345,6 +346,7 @@ class Neighbors:
     def __init__(
         self,
         adata: AnnData,
+        *,
         n_dcs: int | None = None,
         neighbors_key: str | None = None,
     ):
```
Comment (ilan-gold) [inline scanpy/neighbors/__init__.py:355]: This is a public API, no? https://scanpy.readthedocs.io/en/stable/generated/scanpy.Neighbors.html Can we not wrap this?

File: scanpy/plotting/_anndata.py [hunk 6]
```diff
@@ -881,24 +933,24 @@ def violin(
...
-            return g
-        elif len(axs) == 1:
-            return axs[0]
-        else:
-            return axs
+    if show:
+        return None
+    if multi_panel and groupby is None and len(ys) == 1:
+        return g
+    return axs[0] if len(axs) == 1 else axs
```
Comment (ilan-gold) [inline scanpy/plotting/_anndata.py]: nit: I would prefer to just see another if block.  don't even need else because it's a guard condition

File: scanpy/preprocessing/_simple.py [hunk 7]
```diff
@@ -289,15 +300,15 @@ def filter_genes(
 
 @singledispatch
 def log1p(
-    X: AnnData | np.ndarray | spmatrix,
+    data: AnnData | np.ndarray | spmatrix,
```
Comment (ilan-gold) [inline scanpy/preprocessing/_simple.py:311]: Again, seems like a breaking change.

File: scanpy/tools/_utils_clustering.py [hunk 8]
```diff
@@ -1,18 +1,41 @@
...
+
 
 def rename_groups(
-    adata, key_added, restrict_key, restrict_categories, restrict_indices, groups
-):
-    key_added = restrict_key + "_R" if key_added is None else key_added
+    adata: AnnData,
+    restrict_key: str,
+    *,
+    key_added: str | None,
```
Comment (ilan-gold) [inline scanpy/tools/_utils_clustering.py:19]: Did you mean to change the order here?  `key_added` is now after `restrict_key`

File: scanpy/tests/test_package_structure.py [hunk 9]
```diff
@@ -1,23 +1,54 @@
...
+    "sc.tl",
+    "sc.pl",
+    "sc.experimental.pp",
+    "sc.external.pp",
+    "sc.external.tl",
+    "sc.external.pl",
+    "sc.external.exporting",
+    "sc.get",
+    "sc.logging",
+    # "sc.neighbors",  # Not documented
```
Comment (ilan-gold) [inline scanpy/tests/test_package_structure.py:35]: Why is this here?

File: scanpy/preprocessing/_simple.py (hunk 7 above)
Comment (ilan-gold) [reply scanpy/preprocessing/_simple.py:311]: I think you need to do this because this is now a part of the testing suite, but I am not sure this is such a good idea with breaking changes, unless I am missing something.  I have definitely used `X=` syntax before for no other reason than it exists

File: scanpy/tests/test_package_structure.py [hunk 10]
```diff
@@ -60,3 +96,99 @@ def test_function_headers(f):
...
+
+
+copy_sigs: defaultdict[str, ExpectedSig | None] = defaultdict(
+    lambda: ExpectedSig(first_name="adata", copy_default=False, return_ann=None)
+)
+# full exceptions
+copy_sigs["sc.external.tl.phenograph"] = None  # external
+copy_sigs["sc.pp.filter_genes_dispersion"] = None  # deprecated
+copy_sigs["sc.pp.filter_cells"] = None  # unclear `inplace` situation
+copy_sigs["sc.pp.filter_genes"] = None  # unclear `inplace` situation
```
Comment (ilan-gold) [inline scanpy/tests/test_package_structure.py:130]: What do you mean by "unclear `inplace` situation?"

**Review by ilan-gold — COMMENTED** [review]
Nothing blocking, but I am pretty worried about the breaking changes.  This isn't really my call though.

File: scanpy/external/pp/_bbknn.py (hunk 1 above)
Comment (flying-sheep) [reply scanpy/external/pp/_bbknn.py]: It changed before this PR. `bbknn` isn’t actually tested, I just saw that things changed when running tests locally.

File: scanpy/external/pp/_hashsolo.py (hunk 2 above)
Comment (flying-sheep) [reply scanpy/external/pp/_hashsolo.py:275]: `list`, `set`, and `dict` are almost never correct in a *parameter* type annotation (use them in return types). One should use `collections.abc.*` instead to express the interface the function actually supports here.

In this particular case, a correct type annotation is impossible: “`collections.abc.Sequence[float]` of length 3” can’t be expressed, so options capturing part of the truth are `collections.abc.Sequence[float]` and `tuple[float, float, float]`.

It’s common practice to pass short fixed-length sequences as tuples, so using the type I did encourages that.

File: scanpy/external/tl/_phenograph.py (hunk 3 above)
Comment (flying-sheep) [reply scanpy/external/tl/_phenograph.py:42]: Technically yes. Practically, people rather use too many positional parameters than too many positional ones.

Maybe adding a `/` after `data` to enforce positional-only parameter specification would be nice in some cases?

File: scanpy/external/tl/_phenograph.py (hunk 4 above)
Comment (flying-sheep) [reply scanpy/external/tl/_phenograph.py]: I use asserts like this just for type assertions. But good point, I broke the logic here, and since we don’t test external APIs, this would break.

File: scanpy/tools/_utils_clustering.py (hunk 8 above)
Comment (flying-sheep) [reply scanpy/tools/_utils_clustering.py:19]: yes, I did, because it’s derived. I updated all call sites.

File: scanpy/tests/test_package_structure.py (hunk 10 above)
Comment (flying-sheep) [reply scanpy/tests/test_package_structure.py:130]: The `filter_cells` and `filter_genes` functions for some reason say that `copy` is deprecated despite us deciding that `inplace` should be deprecated, so that’s confusing.

File: scanpy/tests/test_package_structure.py (hunk 9 above)
Comment (flying-sheep) [reply scanpy/tests/test_package_structure.py:35]: This module is not part of the public API despite having a name not starting with `_`, so I mention that here in case people go over the public submodules and wonder.

File: scanpy/external/tl/_phenograph.py (hunk 3 above)
Comment (flying-sheep) [reply scanpy/external/tl/_phenograph.py:44]: see other one.

File: scanpy/preprocessing/_simple.py (hunk 7 above)
Comment (flying-sheep) [reply scanpy/preprocessing/_simple.py:311]: Hmm, I guess we can add a helper for those cases. (add the old name as kw-only arg and throw deprecation warning if specified that way)

File: scanpy/tests/test_package_structure.py (hunk 9 above)
Comment (ilan-gold) [reply scanpy/tests/test_package_structure.py:35]: Got it.

File: scanpy/external/pp/_hashsolo.py (hunk 2 above)
Comment (ilan-gold) [reply scanpy/external/pp/_hashsolo.py:275]: Great, wanted to understand best practices yeah

File: scanpy/preprocessing/_simple.py (hunk 7 above)
Comment (ilan-gold) [reply scanpy/preprocessing/_simple.py:311]: Sounds good!
